### PR TITLE
Support Nano Ledger X

### LINKED ISSUE #174: Ledger Nano X
The device is detected by `enumerate` if you set `LEDGER_DEVICE_ID = 0x0004`. I'm able to get keys from it and display an address. Will try some other things later and make a PR.

### PR DESCRIPTION
**Pull request by Sjors**
Fixes #174.

It now detects both model S and model X, which is just a different device number. No need to update any drivers just yet, and there's no need to split `devices/ledger.py` for now.

I didn't test multisignature and advanced script types. Although I suspect it will work identically to the Nano S, I marked them with `?` in the overview.

I only tested via USB, not BlueTooth.

I noticed a bug where if the device falls asleep - it suffers from narcolepsy compared to the model S - and wakes up again, it won't shop anymore until you quit the Bitcoin (testnet) app and enter the app again. `enumerate` will think it's still asleep.

Being able to see a bech32 address on screen without having to scroll is awesome though.

### CODE REVIEW

File: test/test_ledger.py [hunk 1]
```diff
@@ -34,41 +34,41 @@ def test_pin(self):
...
             self.assertIn('code', result)
-            self.assertEqual(result['error'], 'The Ledger Nano S does not need a PIN sent from the host')
+            self.assertEqual(result['error'], 'The Ledger Nano S and X do not need a PIN sent from the host')
             self.assertEqual(result['code'], -9)
 
             result = self.do_command(self.dev_args + ['sendpin', '1234'])
             self.assertIn('error', result)
             self.assertIn('code', result)
-            self.assertEqual(result['error'], 'The Ledger Nano S does not need a PIN sent from the host')
+            self.assertEqual(result['error'], 'The Ledger Nano S and X do not  need a PIN sent from the host')
```
Comment (jonatack) [inline test/test_ledger.py]: nits: extra space after "not" in every message from here on down. Necessary?

File: docs/ledger.md [hunk 2]
```diff
@@ -17,7 +21,7 @@ Currently implemented commands:
 
 ## `signtx` Caveats
 
-Due to device limitiations, not all kinds of transactions can be signed by a Ledger. 
+Due to device limitiations, not all kinds of transactions can be signed by a Ledger.
```
Comment (jonatack) [inline docs/ledger.md]: nit: limitations

File: hwilib/devices/ledger.py [hunk 3]
```diff
@@ -308,53 +311,54 @@ def display_address(self, keypath, p2sh_p2wpkh, bech32):
...
 
     # Prompt pin
     def prompt_pin(self):
-        raise UnavailableActionError('The Ledger Nano S does not need a PIN sent from the host')
+        raise UnavailableActionError('The Ledger Nano S and X do not need a PIN sent from the host')
 
     # Send pin
     def send_pin(self, pin):
-        raise UnavailableActionError('The Ledger Nano S does not need a PIN sent from the host')
+        raise UnavailableActionError('The Ledger Nano S and X does not need a PIN sent from the host')
```
Comment (jonatack) [inline hwilib/devices/ledger.py]: nit: "do not"?

**Comment by Sjors** [comment]
@jonatack thanks, nits fixed.

**Comment by jonatack** [comment]
:racing_car: :dash:

**Comment by instagibbs** [comment]
> I didn't test multisignature and advanced script types

You can run the CI tests manually using Ledger, which test everything we support with Nano S.

**Comment by Sjors** [comment]
`test_signtx` displays a bunch of transactions on my device which I've all approved, but it fails.

```
======================================================================
ERROR: ledger: test_signtx (test_device.TestSignTx)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bitcoin/dev/HWI/test/test_device.py", line 411, in test_signtx
    self._test_signtx("legacy", self.type in supports_multisig)
  File "/Users/bitcoin/dev/HWI/test/test_device.py", line 399, in _test_signtx
    self._generate_and_finalize(True, psbt)
  File "/Users/bitcoin/dev/HWI/test/test_device.py", line 308, in _generate_and_finalize
    self.assertTrue(single_input == self.wrpc.finalizepsbt(first_sign_res['psbt'])['complete'])
  File "/Users/bitcoin/Library/Caches/pypoetry/virtualenvs/HWI-py3.6/lib/python3.6/site-packages/bitcoinrpc/authproxy.py", line 139, in __call__
    response = self._get_response()
  File "/Users/bitcoin/Library/Caches/pypoetry/virtualenvs/HWI-py3.6/lib/python3.6/site-packages/bitcoinrpc/authproxy.py", line 179, in _get_response
    http_response = self.__conn.getresponse()
  File "/Users/bitcoin/.pyenv/versions/3.6.8/lib/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/Users/bitcoin/.pyenv/versions/3.6.8/lib/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/Users/bitcoin/.pyenv/versions/3.6.8/lib/python3.6/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/Users/bitcoin/.pyenv/versions/3.6.8/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 54] Connection reset by peer

----------------------------------------------------------------------
```

I had to manually disable the test for big transactions. Otherwise it fails and the others tests fail along with it. While `ledger: test_big_tx` is running, nothing is displayed on the device.

```
ledger: test_getkeypool (test_device.TestGetKeypool) ... ok
ledger: test_getkeypool_bad_args (test_device.TestGetKeypool) ... ok
ledger: test_big_tx (test_device.TestSignTx) ... FAIL
ledger: test_signtx (test_device.TestSignTx) ... ERROR
ledger: test_display_address_bad_args (test_device.TestDisplayAddress) ... FAIL
ledger: test_display_address_bad_path (test_device.TestDisplayAddress) ... FAIL
ledger: test_display_address_descriptor (test_device.TestDisplayAddress) ... ERROR
ledger: test_display_address_path (test_device.TestDisplayAddress) ... ok
ledger: test_bad_path (test_device.TestSignMessage) ... FAIL
ledger: test_sign_msg (test_device.TestSignMessage) ... ok

======================================================================
ERROR: ledger: test_signtx (test_device.TestSignTx)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bitcoin/dev/HWI/test/test_device.py", line 411, in test_signtx
    self._test_signtx("legacy", self.type in supports_multisig)
  File "/Users/bitcoin/dev/HWI/test/test_device.py", line 328, in _test_signtx
    import_result = self.wrpc.importmulti(keypool_desc)
  File "/Users/bitcoin/Library/Caches/pypoetry/virtualenvs/HWI-py3.6/lib/python3.6/site-packages/bitcoinrpc/authproxy.py", line 141, in __call__
    raise JSONRPCException(response['error'])
bitcoinrpc.authproxy.JSONRPCException: -3: Expected type array, got object
``` 

I don't have time to debug this at the moment...

File: test/test_ledger.py [hunk 4]
```diff
@@ -34,41 +34,41 @@ def test_pin(self):
             result = self.do_command(self.dev_args + ['promptpin'])
             self.assertIn('error', result)
             self.assertIn('code', result)
-            self.assertEqual(result['error'], 'The Ledger Nano S does not need a PIN sent from the host')
+            self.assertEqual(result['error'], 'The Ledger Nano S and X do not need a PIN sent from the host')
```
Comment (nopara73) [inline test/test_ledger.py]: Can you try to never modify the errors? It's a CLI software so software that depends on it relies on parsing error strings.

Comment (nopara73) [reply test/test_ledger.py]: These particularly are not problematic for us, but it may be for others.

Comment (nopara73) [reply test/test_ledger.py]: I mean it's a terrible precedent. Please keep this in mind in further PRs: https://lkml.org/lkml/2012/12/23/75

> If a change results in user programs breaking, it's a bug in the
kernel. We never EVER blame the user programs. How hard can this be to
understand?

Comment (jonatack) [reply test/test_ledger.py]: Hmmm. Errors would ideally return error numbers as well and warn devs to not depend on error message strings.

Comment (nopara73) [reply test/test_ledger.py]: > Hmmm. Errors would ideally return error numbers as well and warn devs to not depend on error message strings.

Yes, indeed that's the clean way to do this: https://github.com/bitcoin-core/HWI/issues/180

It however only partly addresses the more general issue when not error messages, are broken. That being said I may be overly cautious here and maybe the maintainers are experienced enough with maintaining other software that builds on top of other software and libraries to understand that breaking even the smallest things are extremely dangerous and can result in far reaching consequences. This is even more true when we talk about a software that handles money.

Comment (instagibbs) [reply test/test_ledger.py]: > If a change results in user programs breaking

Ugh no, people really shouldn't be relying on exact error messages. If error messages become stale that means they are invalid for manual consumption and users will be unable to fix their issue. Error codes please.

Comment (achow101) [reply test/test_ledger.py]: Are you all blind or have you not tested HWI yourself? Literally every error has an associated error code already. The existence of the an error code is tested in the line directly above this one, and the code itself is tested in the line after this.

Comment (nopara73) [reply test/test_ledger.py]: I see you have error codes on normal commands, however you don't on `enumerate`. I'm only parsing enumerate errors for example, that's why I didn't see them before.

```
[{"type": "coldcard", "path": "...", "error": "Could not open client or get fingerprint information: 'NoneType' object is not callable"}]
```

**Comment by instagibbs** [comment]
needs rebase

**Comment by Sjors** [comment]
Blind rebase (didn't test on a device yet)

**Comment by Sjors** [comment]
Rebased again. Manually tested extracting keys (also with #137) and displaying an address.

**Comment by instagibbs** [comment]
What's needed for merge? Without a simulator I think we basically need to hunt down another person with an X, or just optimistically merge. I could test that you didn't break nano s...

**Comment by achow101** [comment]
I suppose the plan is to just optimistically merge this, so long as Nano S support isn't broken.

I'll try to get a Nano X to test with.

**Comment by Sjors** [comment]
Basic sending and receiving works for me on both X and S with this PR, but I haven't done anything super fancy.

**Comment by instagibbs** [comment]
needs rebase again

```
<provoostenator> Did we improve the test suite so it tells us which button to press?
<provoostenator> Because afaik you can't just hit OK on the device all the time
<instagibbs> yeah you can?>
<instagibbs> at least last time i ran it it did
<instagibbs> test_big_tx takes foreveeeeeeer
<instagibbs> it's a lot of clicking of course
<provoostenator> Oh ok, I thought it also tested a rejection somewhere. maybe not
<provoostenator> When I ran it some of the tests didn't pass on the X.
<instagibbs> oh, that's a self-NACK then :)
```

It passes Nano S tests at least :)

**Comment by promag** [comment]
I have a x, I can test this weekend.

**Comment by instagibbs** [comment]
simple rebase onto master

**Comment by Sjors** [comment]
Added device specific test command: `./run_tests.py --ledger-x`

`ledger: test_signtx (test_device.TestSignTx) ... ERROR` seems consistent. I approved about 3 transactions on the device. The sign message tests blows up after I approve the hash:
```
ERROR: ledger: test_signtx (test_device.TestSignTx)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sjors/dev/HWI/test/test_device.py", line 395, in test_signtx
    self._test_signtx("legacy", self.full_type in supports_multisig)
  File "/Users/sjors/dev/HWI/test/test_device.py", line 380, in _test_signtx
    psbt = self.wrpc.walletcreatefundedpsbt([], [{self.wpk_rpc.getnewaddress('', 'legacy'):(i+1)*out_amt}, {self.wpk_rpc.getnewaddress('', 'p2sh-segwit'):(i+1)*out_amt}, {self.wpk_rpc.getnewaddress('', 'bech32'):(i+1)*out_amt}], 0, {'includeWatching': True, 'subtractFeeFromOutputs': [0, 1, 2]}, True)
  File "/Users/sjors/.pyenv/versions/3.6.8/lib/python3.6/site-packages/bitcoinrpc/authproxy.py", line 139, in __call__
    response = self._get_response()
  File "/Users/sjors/.pyenv/versions/3.6.8/lib/python3.6/site-packages/bitcoinrpc/authproxy.py", line 179, in _get_response
    http_response = self.__conn.getresponse()
  File "/Users/sjors/.pyenv/versions/3.6.8/lib/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/Users/sjors/.pyenv/versions/3.6.8/lib/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/Users/sjors/.pyenv/versions/3.6.8/lib/python3.6/http/client.py", line 266, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response
```

File: test/test_ledger.py [hunk 5]
```diff
@@ -34,48 +34,48 @@ def test_pin(self):
...
-            self.assertEqual(result['error'], 'The Ledger Nano S does not support creating a backup via software')
+            self.assertEqual(result['error'], 'The Ledger Nano S and X do not support creating a backup via software')
             self.assertEqual(result['code'], -9)
 
     # Generic Device tests
     suite = unittest.TestSuite()
     suite.addTest(DeviceTestCase.parameterize(TestLedgerDisabledCommands, rpc, userpass, 'ledger', 'ledger', path, fingerprint, master_xpub, interface=interface))
     suite.addTest(DeviceTestCase.parameterize(TestDeviceConnect, rpc, userpass, 'ledger', 'ledger', path, fingerprint, master_xpub, interface=interface))
-    suite.addTest(DeviceTestCase.parameterize(TestDeviceConnect, rpc, userpass, 'ledger_nano_s', 'ledger', path, fingerprint, master_xpub, interface=interface))
+    suite.addTest(DeviceTestCase.parameterize(TestDeviceConnect, rpc, userpass, device_model, 'ledger', path, fingerprint, master_xpub, interface=interface))
```
[the remark this answers is not in the document]
Comment (Sjors) [reply test/test_ledger.py]: I think I replicated the Trezor variant, but @achow101 will know...

**Comment by instagibbs** [comment]
needs rebase

File: test/test_ledger.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sjors) [reply test/test_ledger.py]: `TestDeviceConnect` checks the exact device model, which is why you need to run the test suite with `--ledger-s` or `--ledger-x`. At the moment that's the only difference, but that might change.

File: test/test_ledger.py [hunk 6]
```diff
@@ -14,7 +14,7 @@
 
 from hwilib.cli import process_commands
 
-def ledger_test_suite(rpc, userpass, interface):
+def ledger_test_suite(device_model, rpc, userpass, interface):
```
[the remark this answers is not in the document]
Comment (Sjors) [reply test/test_ledger.py]: Fixed, you now have to use `test/test_ledger.py --nano-s`

**Comment by Sjors** [comment]
Rebased. To run the Nano X tests:

```
test/test_ledger.py path_to_bitcoind --nano-x
```

File: test/test_ledger.py [hunk 7]
```diff
@@ -86,10 +86,16 @@ def test_backup(self):
     parser = argparse.ArgumentParser(description='Test Ledger implementation')
     parser.add_argument('bitcoind', help='Path to bitcoind binary')
     parser.add_argument('--interface', help='Which interface to send commands over', choices=['library', 'cli', 'bindist'], default='library')
+    model_group = parser.add_mutually_exclusive_group(required=True)
```
[the remark this answers is not in the document]
Comment (Sjors) [reply test/test_ledger.py]: Done

File: test/test_ledger.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sjors) [reply test/test_ledger.py]: Done

**Comment by Sjors** [comment]
Slightly changed the syntax:

```sh
./test/test_ledger.py ../bitcoin/src/bitcoind ledger_nano_s
```

**Comment by instagibbs** [comment]
Last issue seems fixed with https://github.com/bitcoin-core/HWI/pull/242

**Comment by achow101** [comment]
ACK dc0181d2d7c691077f181aa954f665b5865f1675

**Comment by instagibbs** [comment]
ACK

**Pull request merged**